File: Algoritmos/C++/Grafos/Articulation Points.cpp
int n; 
vector<vector<int>> adj;
vector<bool> visited;
vector<int> d, low;
vector<int> ap;
int time;
void dfs(int v, int p = -1) {
    visited[v] = true;
    d[v] = low[v] = time++;
    int children=0;
    for (int to : adj[v]) {
        if (to == p) continue;
        if (visited[to]) {
            low[v] = min(low[v], d[to]);
        } else {
            dfs(to, v);
            low[v] = min(low[v], low[to]);
            if (low[to] >= d[v] && p!=-1)
                ap.push_back(v);
            ++children;
        }
    }
    if(p == -1 && children > 1)
        ap.push_back(v);
}
void find_cutpoints() {
    time = 0;
    visited.assign(n, false);
    d.assign(n, -1);
    low.assign(n, -1);
    for (int i = 0; i < n; ++i) {
        if (!visited[i])
            dfs (i);
    }
}

File: Algoritmos/Java/Grafos/Articulation Points.java
static int n;
static ArrayList<Integer>[] adj;
static int[] d, low;
static int time;
static boolean[] ap;
static void dfs(int u, int par) {
    d[u] = time++;
    low[u] = d[u];
    int children = 0;
    for(int v : adj[u]) {
        if (v == par) continue;
        else if (d[v] > 0)
            low[u] = Math.min(low[u], d[v]);
        else {
            children++;
            dfs(v, u);
            low[u] = Math.min(low[u], low[v]);
            if (par != -1 && low[v] >= d[u]) ap[u] = true;
        }
    }
    if (par == -1 && children > 1) ap[u] = true;
}

static void get_articulation_points() {
    d = new int[n];
    low = new int[n];
    time = 1;
    ap = new boolean[n];
    Arrays.fill(ap, false);
    for (int i = 0; i < n; i++)
        if (d[i] == 0) dfs(i, -1);
}
